File: 240819/경험치를 빠르게 얻기/gain-exp-quickly.cpp
#include <iostream>
#include <algorithm>
#include <climits>
using namespace std;
#define MAX_N 100
#define MAX_T 10000

int n, m, t;
int exp[MAX_N+1], runtime[MAX_N+1];
int dp[MAX_N+1][MAX_T+1];

void Init() {
    for(int i=0; i<=n; i++) {
        for(int j=0; j<=t; j++) {
            dp[i][j] = INT_MIN;
        }
    }

    dp[0][0] = 0;
}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n>>m;
    for(int i=1; i<=n; i++) {
        cin >> exp[i]>>runtime[i];
        t += runtime[i];
    }

    Init();

    for(int i=1; i<=n; i++) {
        for(int j=0; j<=t; j++) {
            // 퀘스트를 진행했을 때
            if(j-runtime[i] >= 0)
                dp[i][j] = max(dp[i][j], dp[i-1][j-runtime[i]]+exp[i]);
            
            // 퀘스트를 진행하지 않았을 때
            dp[i][j] = max(dp[i][j], dp[i-1][j]);
        }
    }

    int ans = INT_MAX;
    for(int j=0; j<=t; j++) {
        if(dp[n][j] >= m)
            ans = min(ans, j);
    }

    if(ans == INT_MAX)
        ans = -1;

    cout << ans;

    return 0;
}

File: 250527/경험치를 빠르게 얻기/gain-exp-quickly.py
import sys

INT_MAX = sys.maxsize
INT_MIN = -sys.maxsize


n, m = tuple(map(int, input().split()))

qest = [(0, 0)] + [
    tuple(map(int, input().split()))
    for _ in range(n)
]

t = sum(time for _ , time in qest)

dp = [[INT_MIN]*(t+1) for _ in range(n+1)]

dp[0][0] = 0

for i in range(1, n+1):
    exp, runtime = qest[i]

    for j in range(t+1):
        if j >= runtime:
            dp[i][j] = max(dp[i][j], dp[i-1][j-runtime]+exp)
        
        dp[i][j] = max(dp[i][j], dp[i-1][j])

ans = INT_MAX

for j in range(t+1):
    if dp[n][j] >= m:
        ans = min(ans, j)

if ans == INT_MAX:
    ans = -1

print(ans)